src/core/mount.c

```diff
@@ -1387,6 +1387,141 @@ static int mount_dispatch_timer(sd_event_source *source, usec_t usec, void *user
         return 0;
 }
 
+typedef struct {
+        bool is_mounted;
+        bool just_mounted;
+        bool just_changed;
+} MountSetupFlags;
+
+static int mount_setup_new_unit(
+                Unit *u,
+                const char *what,
+                const char *where,
+                const char *options,
+                const char *fstype,
+                MountSetupFlags *flags) {
+
+        MountParameters *p;
+        int r;
+
+        assert(u);
+        assert(flags);
+
+        MOUNT(u)->where = strdup(where);
+        if (!MOUNT(u)->where) {
+                r = -ENOMEM;
+                goto fail;
+        }
+
+        u->source_path = strdup("/proc/self/mountinfo");
+        if (!u->source_path) {
+                r = -ENOMEM;
+                goto fail;
+        }
+
+        /* Make sure to initialize those fields before mount_is_extrinsic(). */
+        MOUNT(u)->from_proc_self_mountinfo = true;
+        p = &MOUNT(u)->parameters_proc_self_mountinfo;
+
+        p->what = strdup(what);
+        p->options = strdup(options);
+        p->fstype = strdup(fstype);
+        if (!p->what || !p->options || !p->fstype) {
+                r = -ENOMEM;
+                goto fail;
```

Comment:
Just do `return -ENOMEM`, no need for the `goto`.

Also, you could move the intialization of `source_path` here too. Keeping it separate just makes the code longer.